### PR TITLE
Elasticsearch support

### LINKED ISSUE #411: Add support to create Elasticsearch Domains
Cloudformation recently has implemented the creation of Elasticsearch domains, would be nice to have this feature in troposphere too.

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticsearch-domain.html

### PR DESCRIPTION
**Pull request by edubxb**
Implemented AWS::Elasticsearch resource creation.

Fixes #411

### CODE REVIEW

**Comment by phobologic** [comment]
Hey @edubxb, thanks for this.  Can you provide an example in `examples` and the output in `tests/examples_output`?  That way this will be tested, and it'll give folks an idea of how to use it.

A good example would probably be this: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticsearch-domain.html#d0e51584

Thanks!

**Comment by edubxb** [comment]
Yeah, working on it!

Also, I'm improving the code a bit, with more input values validations.

Sorry for the late response.

**Comment by ink-adavison** [comment]
I'd recommend allowing for a bit of flexibility with the validations. Currently the EBSOptions IOPS validation prevents using Cloudformation Condition functions to configure the IOPS volume type:
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html

e.g. The validation currently prevents something along the lines of:

    par_iops = Parameter(
        'IOPS',
        Type='Number',
        Description='Dedicated IOPS for the volumes (zero for disabled)',
    )

...

    con_iops = ['con_iops',
                Not(Equals(`
                Ref(par_iops), "0"))]

...

    EBSOptions(
        "EBSOptions",
        EBSEnabled=True,
        Iops=If('con_iops', Ref(par_iops), Ref("AWS::NoValue")),
        VolumeSize=Ref(par_volume_size),
        VolumeType=If('con_iops', 'io1', 'gp2')
    )

**Comment by phobologic** [comment]
@ink-adavison can you provide the exception output?  `If` should be a `AWSHelperFn`, which should be allowed (and un-validated) by troposphere for any field because of this exact issue.

**Comment by ink-adavison** [comment]
Sure, here's an example to try:
https://gist.github.com/ink-adavison/20562a3fb898babb93cc

I got the same error when using simple Ref functions instead of Ifs too.

The exception is:

    Traceback (most recent call last):
      File "examples/Elasticsearch.py", line 69, in <module>
        main()
      File "examples/Elasticsearch.py", line 66, in main
        print(template.to_json())
      File "build/bdist.linux-x86_64/egg/troposphere/__init__.py", line 470, in to_json
      File "/usr/lib/python2.7/json/__init__.py", line 250, in dumps
        sort_keys=sort_keys, **kw).encode(obj)
      File "/usr/lib/python2.7/json/encoder.py", line 209, in encode
        chunks = list(chunks)
      File "/usr/lib/python2.7/json/encoder.py", line 434, in _iterencode
        for chunk in _iterencode_dict(o, _current_indent_level):
      File "/usr/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
        for chunk in chunks:
      File "/usr/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
        for chunk in chunks:
      File "/usr/lib/python2.7/json/encoder.py", line 443, in _iterencode
        for chunk in _iterencode(o, _current_indent_level):
      File "/usr/lib/python2.7/json/encoder.py", line 434, in _iterencode
        for chunk in _iterencode_dict(o, _current_indent_level):
      File "/usr/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
        for chunk in chunks:
      File "/usr/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
        for chunk in chunks:
      File "/usr/lib/python2.7/json/encoder.py", line 442, in _iterencode
        o = _default(o)
      File "build/bdist.linux-x86_64/egg/troposphere/__init__.py", line 367, in default
      File "build/bdist.linux-x86_64/egg/troposphere/__init__.py", line 176, in JSONrepr
      File "build/bdist.linux-x86_64/egg/troposphere/elasticsearch.py", line 46, in validate
    ValueError: Cannot specify Iops if VolumeType is not 'io1'.

**Comment by phobologic** [comment]
Ahh, right - I forgot that this PR overrides validate() to provide it's own validation on top of the standard stuff.  Thanks @ink-adavison !

**Comment by edubxb** [comment]
Fixed the issue when using Conditionals.

**Comment by jdilley** [comment]
@phobologic is this good? I'm actually anxiously waiting for this change to be available. Thanks @edubxb !

**Comment by edubxb** [comment]
I've done some checks, but any additional feedback will be wellcome!

File: troposphere/elasticsearch.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+try:
+    from awacs.aws import Policy
+    policytypes = (dict, Policy)
+except ImportError:
+    policytypes = dict,
+
+
+def validate_volume_type(volume_type):
+    """Validate VolumeType for ElasticsearchDomain"""
+    if isinstance(volume_type, AWSHelperFn):
```
Comment (phobologic) [inline troposphere/elasticsearch.py]: You shouldn't need this, due to this:

https://github.com/cloudtools/troposphere/blob/master/troposphere/__init__.py#L99

**Comment by phobologic** [comment]
This is great - thanks for all the hard work on this @edubxb - just the one comment, then I'll merge it.  Sorry for the delay - work got crazy!

File: troposphere/elasticsearch.py (hunk 1 above)
Comment (edubxb) [inline troposphere/elasticsearch.py]: Fixed!

**Comment by edubxb** [comment]
I also removed the unnecessary Tag Class that I "borrowed" from the ec2 module.

**Comment by phobologic** [comment]
Awesome!  Thanks @edubxb !

**Pull request merged**